components/server/src/workspace/context-parser-service.ts

```diff
@@ -70,6 +63,88 @@ export class ContextParser {
+        const span = TraceContext.startSpan("ContextParser.internalHandle", ctx);
+        let result: WorkspaceContext | undefined;
+
+        for (const parser of this.allContextParsers) {
+            if (parser.canHandle(user, nonPrefixedContextURL)) {
+                result = await parser.handle({ span }, user, nonPrefixedContextURL);
+                break;
+            }
+        }
+        if (!result) {
+            throw new Error(`Couldn't parse context '${nonPrefixedContextURL}'.`);
+        }
+
+        // TODO: Make the parsers return the context with normalizedContextURL set
+        result.normalizedContextURL = nonPrefixedContextURL;
+        return result;
+    }
+
+    protected buildUpstreamCloneUrl(context: CommitContext): string | undefined {
+        let upstreamCloneUrl: string | undefined = undefined;
+        if (PullRequestContext.is(context) && context.base) {
+            upstreamCloneUrl = context.base.repository.cloneUrl;
+        } else if (context.repository.fork) {
+            upstreamCloneUrl = context.repository.fork.parent.cloneUrl;
+        }
+
+        if (context.repository.cloneUrl === upstreamCloneUrl) {
+            return undefined;
+        }
+        return upstreamCloneUrl;
+    }
+
+    protected async handleMultiRepositoryContext(ctx: TraceContext, user: User, context: WorkspaceContext): Promise<WorkspaceContext> {
+        if (!CommitContext.is(context)) {
+            return context;
+        }
+        const span = TraceContext.startSpan("ContextParser.handleMultiRepositoryContext", ctx);
+        let config = await this.configProvider.fetchConfig({ span }, user, context);
+        let mainRepoContext: WorkspaceContext | undefined;
+        if (config.config.mainConfiguration) {
+            mainRepoContext = await this.internalHandleWithoutPrefix({ span }, user, config.config.mainConfiguration);
+            if (!CommitContext.is(mainRepoContext)) {
+                throw new InvalidGitpodYMLError([`Cannot find main repository '${config.config.mainConfiguration}'.`]);
+            }
+            config = await this.configProvider.fetchConfig({ span }, user, mainRepoContext);
+        }
+
+        if (config.config.additionalRepositories && config.config.additionalRepositories.length > 0) {
+            const subRepoCommits: GitCheckoutInfo[] = [];
+            for (const subRepo of config.config.additionalRepositories) {
+                let subContext = await this.internalHandleWithoutPrefix({ span }, user, subRepo.url) as CommitContext;
+                if (!CommitContext.is(subContext)) {
+                    throw new InvalidGitpodYMLError([`Cannot find sub-repository '${subRepo.url}'.`]);
+                }
+                if (context.repository.cloneUrl === subContext.repository.cloneUrl) {
+                    // if it's the repo from the original context we want to use that commit.
+                    subContext = JSON.parse(JSON.stringify(context));
+                }
```

Comment:
Is there any point in allowing a repository to specify itself as an additional repository? 🤔 (And then fetch the same commit history multiple times etc?) Naively, I would want to just "skip" this case:

```suggestion
                if (context.repository.cloneUrl === subContext.repository.cloneUrl) {
                    continue;
                }
```

Reply:
The case is not, that someone created a self-pointer but that the context URL points to the main repo.